ethereum_trie/utils/binaries.nim

```diff
@@ -1,43 +1,77 @@
-import
-  algorithm, eth_utils, cytoolz, trie.constants
-
-proc decodeFromBin*(inputBin: string): string =
-  ##     0100000101010111010000110100100101001001 -> ASCII
-  newString(result)
-  for chunk in partitionAll(8, inputBin):
-    var sum = 0
-    for exp, bit in reverse(chunk):
-      sum += 2^exp * bit
-
-    result.add(sum)
- 
-proc encodeToBin*(value: string): seq[Byte] =
-  ##     ASCII -> 0100000101010111010000110100100101001001
-  newSeq(result)
-  for c in value:
-    for exp in EXP:
-      if c and exp:
-        result.add('1')
-      else:
-        result.add('0')
-  
-proc encodeFromBinKeypath*(inputBin: string): string =
-  ##     Encodes a sequence of 0s and 1s into tightly packed bytes
-  ##     Used in encoding key path of a KV-NODE
-  var paddedBin = nil
-  var prefix = TWOBITS[nil]
-  if len(paddedBin) mod 8 == 4:
-    return decodeFromBin(nil)
+import rlp/types, strutils
+
+type
+  BinVector* = BytesRange
+
+const
+  binaryZero = byte(0x00)
+  binaryOne = byte(0x01)
+
+func generateMask(): array[8, byte] {.compileTime.} =
+  for i in 0..<8:
+    result[7 - i] = byte(1 shl i)
+
+const
+  byteMask = generateMask()
+  TWO_BITS = ["\x00\x00", "\x00\x01", "\x01\x00", "\x01\x01"]
+  PREFIX_00 = "\x00\x00"
+  PREFIX_100000 = "\x01\x00\x00\x00\x00\x00"
+
+proc encodeToBin*(value: BytesRange): Bytes =
```

Comment:
Isn't it possible to remove these memory allocating functions that store bits as bytes?

We can just implement helpers for accessing the individual bits of a `Bytes` values.